lib/ansible/plugins/connection/__init__.py

```diff
@@ -148,6 +171,58 @@ def transport(self) -> str:
     def _connect(self: T) -> T:
         """Connect to the host we've been initialized with"""
 
+    def prepare_command(
+        self,
+        info: PrepareCommandInfo,
+    ) -> tuple[str, bytes | None, bool, dict[str, t.Any]]:
+        """Prepares a complex command for exec_command.
+
+        This is the base class that implements all the historical rules Ansible
+        has when preparing a command from _low_level_execute_command() before
+        it calls exec_command(). It can be overridden in a connection
+        implementation if requires custom handling or to embed more information
+        passed through to exec_command.
+
+        The return value is a tuple consisting of the following
+
+        * cmd: the cmd str that will be passed to exec_command
+        * in_data: if set, the data to be passed to the command's stdin
+        * use_become: historically known as sudoable, signals whether the cmd
+            has been wrapped in a become command
+        * exec_kwargs: custom kwargs to be passed to exec_command that are
+            specific to the connection plugin
+
+        :arg info: the command information that should be prepared.
+        :returns: a tuple of (cmd, in_data, use_become, exec_kwargs).
+        """
+        cmd = info.cmd
+        if info.chdir:
+            display.debug(f"prepare_command(): changing cwd to '{info.chdir}' for this command")
+            cmd = self._shell.append_command('cd %s' % info.chdir, cmd)
+
+        if info.use_become and self.become:
+            display.debug("prepare_command(): using become for this command")
+            cmd = self.become.build_become_command(cmd, self._shell)
+
+        # FUTURE: allow_executable should be deprecated and connections that
+        # can't use executable handling should use a custom prepare_command
+        # implementation.
+        # Windows is special and doesn't support any of the executable handling.
+        if self.allow_executable and not getattr(self._shell, "_IS_WINDOWS", False):
+            executable = info.executable
+            if executable is None:
+                executable = self._play_context.executable
+
+                # FIXME: Integrate this into ssh only.
+                # mitigation for SSH race which can drop stdout (https://github.com/ansible/ansible/issues/13876)
+                # only applied for the default executable to avoid interfering with the raw action
+                cmd = self._shell.append_command(cmd, 'sleep 0')
+
+            if executable:
+                cmd = executable + ' -c ' + shlex.quote(cmd)
```

Comment:
I'm still not too happy with how this part turned out. The `sleep 0` code should only be set in ssh but for ssh to only do this it would need to reimplement this whole method. Will need to think about this one a bit more which allows a connection plugin to customise this specific bit of code while keeping the cmd + become stuff before this.

Reply:
Not just ssh, paramiko too, this was mostly added for OSs that have a 'immediate success' on fork but really did not work if you closed the connection right away (solaris ...).

Reply:
The code specifically singles out ssh here, even the commit that added it talks about it being a problem for ssh/paramiko https://github.com/ansible/ansible/commit/bad293ae3537a9d951636b571c8b74df6d8a12b0. I wonder if @mattclay remembers the details around this.

Reply:
The `sleep 0` was added specifically to address a Linux kernel bug that affected `sshd` from OpenSSH, see: https://github.com/ansible/ansible/issues/13876#issuecomment-250776339

Since the kernel bug has been fixed for quite a while now, we should consider making this work-around opt-in through config.

Reply:
it also affects solaris due to how they implemented fork